internal/jsonwall/jsonwall.go

```diff
@@ -0,0 +1,307 @@
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sort"
+	"strings"
+)
+
+// DBWriter holds in memory the state of a database while it's being prepared
+// for serialization and implements the WriterTo interface for assembling it.
+type DBWriter struct {
+	options *DBWriterOptions
+	entries [][]byte
+}
+
+type DBWriterOptions struct {
+	// Schema is included in the database header to help the decoding
+	// process. The value is made available when reading, and is not
+	// internally interpreted.
+	Schema string
+}
+
+// NewDBWriter returns a database writer that can assemble new databases.
+func NewDBWriter(options *DBWriterOptions) *DBWriter {
+	if options == nil {
+		options = &DBWriterOptions{}
+	}
+	return &DBWriter{options: options}
+}
+
+// Add encodes the provided value as a JSON object and includes the resulting
+// data into the database being created.
+func (dbw *DBWriter) Add(value any) error {
+	data, err := json.Marshal(value)
+	if err != nil {
+		return err
+	}
+	if len(data) == 0 || data[0] != '{' {
+		return fmt.Errorf("invalid database value: %#v", value)
+	}
+	dbw.entries = append(dbw.entries, data)
+	return nil
+}
+
+type sortableEntries [][]byte
+
+func (e sortableEntries) Less(i, j int) bool { return bytes.Compare(e[i], e[j]) < 0 }
+func (e sortableEntries) Swap(i, j int)      { e[i], e[j] = e[j], e[i] }
+func (e sortableEntries) Len() int           { return len(e) }
+
+type jsonwallHeader struct {
+	Version string `json:"jsonwall"`
+	Schema  string `json:"schema,omitempty"`
+	Count   int    `json:"count"`
+}
+
+const jsonwallVersion = "1.0"
+
+func (dbw *DBWriter) writeHeader(w io.Writer, count int) (int, error) {
+	data, err := json.Marshal(&jsonwallHeader{
+		Version: jsonwallVersion,
+		Schema:  dbw.options.Schema,
+		Count:   count,
+	})
+	if err != nil {
+		return 0, fmt.Errorf("internal error: cannot marshal database header")
+	}
+	return w.Write(append(data, '\n'))
+}
+
+// WriteTo assembles the current database state and writes it to w.
+func (dbw *DBWriter) WriteTo(w io.Writer) (n int64, err error) {
+	m, err := dbw.writeHeader(w, len(dbw.entries)+1)
+	n += int64(m)
+	sort.Sort(sortableEntries(dbw.entries))
+	if err != nil {
+		return n, err
+	}
+	for _, entry := range dbw.entries {
+		m, err := w.Write(entry)
+		n += int64(m)
+		if err == nil {
+			m, err = w.Write([]byte{'\n'})
+			n += int64(m)
+		}
+		if err != nil {
+			return n, err
+		}
+	}
+	return n, nil
+}
+
+// ReadDB reads into memory the database from the provided r.
+func ReadDB(r io.Reader) (*DB, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	record := 0
+	for i := range data {
+		if data[i] == '\n' {
+			record = i + 1
+			break
+		}
+	}
+	var header jsonwallHeader
+	err = json.Unmarshal(data[:record], &header)
+	if err != nil {
+		return nil, fmt.Errorf("invalid database content")
+	}
+	if !strings.HasPrefix(header.Version, jsonwallVersion[:strings.Index(jsonwallVersion, ".")+1]) {
```

Comment:
Without the period, HasPrefix("10.0", "1") is true. We also intentionally don't want to include the minor, because the compatibility is determined by the major version. Anything that is "1.*" should be able to be parsed by a parser that understands version "1.*", otherwise compatibility has been broken and it's not major 1 anymore.

Reply:
Missed that obvious point.